### PR TITLE
fix: capture errors when processing versions response body into promise

### LINKED ISSUE #1: Invalid Valence Version route responses are uncaught
When reading the response body from the valence version route any exception is uncaught. This lead to terminated worker processes in [user-info-api](https://github.com/Brightspace/user-info-api/) and ultimately [downtime](https://github.com/Brightspace/farnsworth-rca/pull/3/commits/3c09e775a0de3a063cf0af195c13f0c13c55043b) in the Pulse Dates Service.

We should [wrap this section](https://github.com/Brightspace/node-valence-url/blob/master/src/ValenceVersions.js#L62) in a try/catch and `reject` the promise similar to what happens if the call to the valence version route simply [fails outright](https://github.com/Brightspace/node-valence-url/blob/master/src/ValenceVersions.js#L58).

### PR DESCRIPTION
**Pull request by omsmith**
fixes #1

### CODE REVIEW

File: src/ValenceVersions.js [hunk 1]
```diff
@@ -49,24 +49,38 @@ function ValenceVersions(opts) {
...
-						productVersions[product.ProductCode] = {
-							latest: product.LatestVersion,
-							supported: product.SupportedVersions.sort().reverse()
-						};
-					});
-
-					resolve(productVersions);
+					resolve(new Promise(function(resolve) {
+						if (err) {
+							throw new errors.UnexpectedVersionsResponse(err);
```
Comment (omsmith) [inline src/ValenceVersions.js]: This part could theoretically be semver-majorish? *shrug*

File: test/ValenceVersions.js [hunk 2]
```diff
@@ -145,6 +145,25 @@ describe('ValenceVersion', function() {
 			lms.done();
 		});
 
+		it('should reject if the response is of an unexpected format', function*() {
+			const lms = nock(tenantUrl)
+				.get('/d2l/api/versions/')
+				.reply(200, 'cats');
```
Comment (omsmith) [inline test/ValenceVersions.js]: Didn't really test all my ifs ands and buts

File: src/ValenceVersions.js [hunk 3]
```diff
@@ -49,24 +49,38 @@ function ValenceVersions(opts) {
...
+						}
+
+						if (!res.body || !Array.isArray(res.body)) {
+							throw new errors.UnexpectedVersionsResponse(new Error(`Repsonse body is not an array. Got "${typeof res.body}"`));
```
Comment (ryantmer) [inline src/ValenceVersions.js]: Would it be helpful here (and other places) to include the response body? Might just end up being enormous log data, granted...

File: src/ValenceVersions.js [hunk 4]
```diff
@@ -49,24 +49,38 @@ function ValenceVersions(opts) {
...
+						}
+
+						const productVersions = {};
+						for (var i = 0; i < res.body.length; ++i) {
```
Comment (ryantmer) [inline src/ValenceVersions.js]: No `forEach` love? 😢

File: src/ValenceVersions.js (hunk 1 above)
Comment (ryantmer) [reply src/ValenceVersions.js]: I guess technically yeah. Might as well!

File: src/ValenceVersions.js (hunk 4 above)
Comment (omsmith) [reply src/ValenceVersions.js]: I'd have to use `some` or `every` in order to short-circuit

**Comment by omsmith** [comment]
Added some variations to the test. Still doesn't cover all of the ifs ands and buts (pretty sure), but does as for as code coverage is concerned apparently

**Pull request merged**